src/libs/blueprint/conduit_blueprint_mesh_topology_metadata.cpp

```diff
@@ -0,0 +1,4111 @@
+    void build_association_3_1_and_3_0_nonph();
+
+    //-----------------------------------------------------------------------
+    /**
+     @param Build a map of edge keys to edge ids. The keys are the hashed endpoints.
+     @param[out] edge_key_to_id A sorted vector of pairs where each pair.first
+                                is the hashed edge key and pair.second is the
+                                edge's global index.
+     */
+    void build_edge_key_to_id(std::vector<std::pair<uint64, index_t>> &edge_key_to_id) const;
+
+    //-----------------------------------------------------------------------
+    /**
+     @brief Look up an edge by key in the map by binary search and return its
+            edge id.
+
+     @param edge_key_to_id The map built by build_edge_key_to_id.
+     @param key            The edge key.
+
+     @return The edge id.
+     */
+    index_t lookup_edge_id(std::vector<std::pair<uint64, index_t>> &edge_key_to_id,
+                           const uint64 key) const;
+
+    //-----------------------------------------------------------------------
+    /**
+     @brief Print association G[e][a].
+     */
+    void print_association(int e, int a, bool global = true) const;
+
+    //-----------------------------------------------------------------------
+    /**
+     @brief This method builds a child to parent association by reversing
+            an existing parent to child association.
+
+     @param e The entity dimension (the thing we have: element=3,face=2,...)
+     @param a The association dimension (the thing we want to know about with
+              respect to the entity dimension). e < a.
+     */
+    void build_child_to_parent_association(int e, int a);
+
+    /**
+      @brief Return all of the data for a GLOBAL association by COPYING it into
+             the provided Conduit node.
+
+      @param src_dim The source dimension of the desired association.
+      @param dst_dim The destination dimension of the desired association.
+      @param[out] map_node The Conduit node that will contain the copied association data.
+
+      @note This method guarantees that all bulk arrays for values, sizes, and
+            offsets will be index_t.
+     */
+    bool get_global_dim_map(index_t src_dim, index_t dst_dim, Node &map_node) const;
+    bool get_local_dim_map(index_t src_dim, index_t dst_dim, Node &map_node) const;
+
+    //-----------------------------------------------------------------------
+    /**
+     @brief Returns the number of entities for L(e,a) so we can iterate through them.
+     @param e The entity dimension
+     @param a The association dimension
+     @return The number of entities in L(e,a).
+     */
+    index_t get_local_association_entity_range(index_t e, index_t a) const;
+
+    //-----------------------------------------------------------------------
+    /**
+     @brief Takes the input topology and reuses it has the highest topology.
```

Comment:
reuses?  ensures?

Reply:
I'll fix: "Takes the input topology and uses it as the highest topology."